File: 02/java/BSTMap.java

public class BSTMap<K extends Comparable<K>, V> implements Map<K, V> {
    private class Node {
        public K key;
        public V value;
		public Node left, right;

		public Node(K key, V value) {
            this.key = key;
            this.value = value;
            this.left = null;
            this.right = null;
		}
    }

    private Node root;
    private int size;

    BSTMap() {
        root = null;
        size = 0;
    }

    private Node getNode(Node node, K key) {
        if( node == null ) return null;

        if( key.compareTo(node.key) == 0 ) {
            return node;
        } else if( key.compareTo(node.key) > 0 ) {
            return getNode(node.right, key);
        } else {
            return getNode(node.left, key);
        }
    }

    @Override
    public int getSize() { return size; }
    
    @Override
    public boolean isEmpty() { return size == 0; }

    @Override
    public boolean contains(K key) { return getNode(root, key) != null; }

    @Override
    public V get(K key) {
        Node node = getNode(root, key);
        return node == null ? null : node.value;
    }

    @Override
    public void set(K key, V value) {
        Node node = getNode(root, key);
        if( node == null ) throw new IllegalArgumentException(key + " doesn't exist!");
        node.value = value;
    }
    
    @Override
    public void add(K key, V value) { root = add(root, key, value); }
	private Node add(Node node, K key, V value) {
		if( node == null ) {
			size ++;
			return new Node(key, value);
		}

		if( key.compareTo(node.key) > 0 ) {
			node.right = add(node.right, key, value);
		} else if( key.compareTo(node.key) < 0 ) {
			node.left = add(node.left, key, value);
		} else {
            node.value = value;
        }

		return node;
	}

    @Override
    public V remove(K key) {
        Node node = getNode(root, key);
        if( node != null ) {
            root = remove(root, key);
            return node.value;
        }
        return null;
    }
    private Node remove(Node node, K key) {
        if( node == null ) return null;

        if( key.compareTo(node.key) > 0 ) {
            node.right = remove(node.right, key);
			return node;
        } else if ( key.compareTo(node.key) < 0 ) {
            node.left = remove(node.left, key);
			return node;
        } else {
            if( node.left == null ) {
				Node rightNode = node.right;
				node.right = null;
				size --;
				return rightNode;
			}

			if( node.right == null ) {
				Node leftNode = node.left;
				node.left = null;
				size --;
				return leftNode;
			}

			Node successor = minimum(node.right);
			successor.right = removeMin(node.right); // 把继任从右子树删除，返回的右子树根节点赋值给继任右节点
			successor.left = node.left;

			node.right = node.left = null;
			return successor;
        }
    }
	private Node minimum(Node node) {
		if( node.left == null ) return node;
		return minimum(node.left);
    }
    private Node removeMin(Node node) {
		if( node.left == null ) {
			Node rightNode = node.right;
			node.right = null;
			size --;
			return rightNode;
		}

		node.left = removeMin(node.left);
		return node;
	}
}

File: 02/js/BSTmap.js

class Node {
    constructor(key, value) {
        this.key = key || null;
        this.value = value || null;
        this.left = null;
        this.right = null;
    }
}

class BSTMap {
    constructor() {
        this.root = null;
        this.size = 0;
    }

    compare(compareFn, key, node) {
        return compareFn ? (
			compareFn(key, node.key)
		) : (
			key === node.key ? 0 : key > node.key ? 1 : -1
        );
    }

    getNode(node, key, compare) {
        if( node == null ) return null;

        const compareRes = this.compare(compare, key, node);
        if( compareRes === 0 ) {
            return node;
        } else if( compareRes > 0 ) {
			return this.getNode(node.right, key, compare);
		} else {
			return this.getNode(node.left, key, compare);
		}
    }

    getSize() { return this.size; }
    isEmpty() { return this.size === 0; }
    contains(key) { return !!this.getNode(this.root, key); }

    get(key, log) { return ((this.getNode(this.root, key) || {}).value || null); }
    set(key, value) {
        const node = this.getNode(this.root, key);
        if( !node ) return console.warn(`${key} doesn't exist!`);
        node.value = value;
    }

    add(key, value, compare) { this.root = this._add(this.root, key, value, compare); }
    _add(node, key, value, compare) {
        if( !node ) {
            this.size ++;
            return new Node(key, value);
        }

        const compareRes = this.compare(compare, key, node);
        if( compareRes > 0 ) {
            node.right = this._add(node.right, key, value, compare);
        } else if ( compareRes < 0 ) {
            node.left = this._add(node.left, key, value, compare);
        } else {
            node.value = value;
        }

        return node;
    }

    remove(key, compare) {
        const node = this.getNode(this.root, key, compare);
        if( node ) {
            root = this._remove(this.root, key, compare);
            return node.value;
        }
        return null;
    }
    _remove(node, key, compare) {
        if( !node ) return null;

        const compareRes = this.compare(compare, key, node);
        if( compareRes > 0 ) {
            node.right = this._remove(node.right, key, compare);
        } else if ( compareRes < 0 ) {
            node.left = this._remove(node.left, key, compare);
        } else {
            if( !node.left ) {
				const rightNode = node.right;
				node.right = null;
				this.size --;
				return rightNode;
			}

			if( !node.right ) {
				const leftNode = node.left;
				node.left = null;
				this.size --;
				return leftNode;
			}

			const successor = minimum(node.right);
			successor.right = removeMin(node.right); // 把继任从右子树删除，返回的右子树根节点赋值给继任右节点
			successor.left = node.left;

            node.right = null;
            node.left = null;
			return successor;
        }
    }
    _minimum(node) {
        if( !node.left ) return node;
        return this.minimum(node.left);
    }
    _removeMin(node) {
        if( !node.left ) {
            const rightNode = node.right;
            node.right = null;
            this.size --;
            return rightNode;
        }

        node.left = this._removeMin(ndoe.left);
        return node;
    }
}

module.exports = BSTMap;
